### PR TITLE
OCPBUGS-31570: Make sure thin provisioning DMs are cleaned correctly

### PR DESCRIPTION
**Pull request by paul-maidment**
When a thin provisioning DM is listed, only the `_tdata` and `_tmeta` parts of that provisioning are listed as associated with a given disk. This results in the disk cleanup trying to delete the `_tdata` and `_tmeta` parts before deleting the base part.

This PR changes that by deriving the base part so that it may be deleted prior to removal of `_tmeta` and `_tdata` parts of the provisioning. This also ensures that the base DMs will always be listed first.

Already tested this by building against master and linking in go.mod of assisted-installer, then using this for a cluster installation in which we manually create a thin provisioning. The provisioning is correctly cleaned and the installation completes, where it failed to do so before.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -3,6 +3,7 @@ module github.com/rh-ecosystem-edge/preinstall-utils
 go 1.20
 
 require (
+	github.com/go-openapi/swag v0.23.0
```
Comment (omertuc) [inline go.mod:6]: We shouldn't have an entire OpenAPI swagger dependency just for swags just for `ContainsString`. If you need a contains function, either bring in the modern `lo` library made for this purpose or use the existing outdated (although already existing) `funk` dependency

**Review by omertuc — CHANGES_REQUESTED** [review]

**Review by omertuc — CHANGES_REQUESTED** [review]
There's still a `swag` dependency

**Comment by paul-maidment** [comment]
> There's still a `swag` dependency

Fixed now

File: go.mod [hunk 2]
```diff
@@ -14,13 +14,10 @@ require (
 require (
 	github.com/fsnotify/fsnotify v1.4.9 // indirect
 	github.com/google/go-cmp v0.6.0 // indirect
-	github.com/kr/text v0.2.0 // indirect
```
Comment (omertuc) [inline go.mod:17]: Why the `go.mod` changes?

**Review by omertuc — CHANGES_REQUESTED** [review]

File: pkg/disk_ops.go [hunk 3]
```diff
@@ -396,13 +397,46 @@ func (r *diskOps) RemoveDMDevice(dmDevice string) error {
...
+
 func (r *diskOps) RemoveAllDMDevicesOnDisk(diskName string) error {
 	var ret error
 	dmDevices, err := r.getDMDevices(diskName)
 	if err != nil {
 		return err
 	}
-	for _, dmDevice := range dmDevices {
+	devices := r.sortDMDevicesForDeletion(append(dmDevices, r.getBaseDMDevicesForThinProvisioning(dmDevices)...))
+	for _, dmDevice := range devices {
```
Comment (omertuc) [inline pkg/disk_ops.go:439]: Why the `dmDevices -> devices` name changes, are the devices in `devices` no longer just dm devices?

**Review by omertuc — CHANGES_REQUESTED** [review]

**Review by omertuc — CHANGES_REQUESTED** [review]
~`go.sum` and `vendor` directory still affected~

EDIT: fixed

File: pkg/disk_ops.go [hunk 4]
```diff
@@ -396,12 +397,45 @@ func (r *diskOps) RemoveDMDevice(dmDevice string) error {
 	return err
 }
 
+// Make sure devices are ordered in a manner that ensures they can be deleted
+// When using thin provisioning we have to contend with _tdata and _tmeta dmsetup records
+// These cannot be deleted until the main dmsetup record is deleted
```
Comment (omertuc) [inline pkg/disk_ops.go:402]: Documentation comment does not follow golang conventions, should begin with the function name / present simple kind of phrasing

**Review by omertuc — CHANGES_REQUESTED** [review]

File: pkg/disk_ops.go [hunk 5]
```diff
@@ -396,12 +397,41 @@ func (r *diskOps) RemoveDMDevice(dmDevice string) error {
 	return err
 }
 
+// sortDMDevicesForDeletion makes sure that DM devices are ordered correctly for deletion.
```
Comment (omertuc) [inline pkg/disk_ops.go:400]: Why did you delete the entire useful paragraph about how they must be deleted in order?

File: pkg/disk_ops.go [hunk 6]
```diff
@@ -396,12 +397,41 @@ func (r *diskOps) RemoveDMDevice(dmDevice string) error {
...
+		if strings.HasSuffix(dmDevice, "_tdata") || strings.HasSuffix(dmDevice, "_tmeta") {
+			dependentDevices = append(dependentDevices, dmDevice)
+			continue
+		}
+		rootDevices = append(rootDevices, dmDevice)
+	}
+	return append(rootDevices, dependentDevices...)
+}
+
+// getBaseDMDevicesForThinProvisioning resolves base DM device names from subdevice names.
```
Comment (omertuc) [inline pkg/disk_ops.go:414]: Here too, the previous documentation was very helpful

File: pkg/disk_ops.go (hunk 5 above)
Comment (paul-maidment) [reply pkg/disk_ops.go:400]: I think the request was to be more terse in the description "present simple kind of phrasing"

**Comment by omertuc** [comment]
We can run CI in the bump PR on the respective Assisted repo
